### PR TITLE
feat(eslint-plugin): [no-unnecessary-boolean-literal-compare] add option to check nullable booleans

### PR DESCRIPTION
**Pull request by zachkirsch**

### CODE REVIEW

**Comment by bradzacher** [comment]
thanks for the PR @zachkirsch - could you please provide some context behind this PR?
Why do you want this change?

To me it seems like this option removes a large portion of the usefulness of the rule as the most common case would be a nullable boolean.

**Comment by zachkirsch** [comment]
Yeah! The motivation for this PR is that I see `myOptionalBoolean !== false` a lot of the time, usually as optional props (that default to `true`) in react components. For example:
```typescript
interface Props {
  // whether to display title (defaults to true)
  displayTitle?: boolean   
}

...

  public render() {
    return (
      <div>
        {this.props.displayTitle !== false && renderTitle()}
        ...
```

Before `??` existed, `myOptionalProp !== false` was the only concise way to check an optional prop and default to `true` if it's `undefined`. In my opinion, I think that `myOptionalProp !== false` is a very confusing way to express "use this prop but default to `true` if it's missing," and IMO it's harder to see bugs that are written that way.

`??` is the perfect replacement for this scenario - it's a built-in operator that provides a default if your variable is undefined at runtime. `(this.props.displayTitle ?? true) && renderTitle()` very clearly means "render the title based on the value of `this.props.displayTitle` (and default to `true` if it's not defined)"

> To me it seems like this option removes a large portion of the usefulness of the rule as the most common case would be a nullable boolean.

I think this rule is still very useful with this option, as it
1. catches unnecessary comparisons with boolean variables
    1. e.g. `myBoolean !== true --> !myBoolean`
1. catches unnecessary comparisons between nullable boolean variables and `true`
    1. e.g. `myOptionalBoolean !== true --> !myOptionalBoolean`
1. Cleans up comparisons between optional booleans and `false` (described above in this comment)

File: packages/eslint-plugin/src/rules/no-unnecessary-boolean-literal-compare.ts [hunk 1]
```diff
@@ -31,18 +47,38 @@ export default util.createRule<[], MessageIds>({
...
+        additionalProperties: false,
+      },
+    ],
     type: 'suggestion',
   },
-  defaultOptions: [],
-  create(context) {
+  defaultOptions: [
+    {
+      allowComparingNullableBooleans: true,
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/no-unnecessary-boolean-literal-compare.ts:72]: for backwards compatibility, please set this to `false` (and update the docs).
I don't think that this is something everyone will want to have on for this rule - as many people like strict boolean expressions.
```suggestion
      allowComparingNullableBooleans: false,
```

File: packages/eslint-plugin/docs/rules/no-unnecessary-boolean-literal-compare.md [hunk 2]
```diff
@@ -36,6 +36,48 @@ if (someUndefinedCondition === false) {
...
+
+```ts
+type Options = {
+  // if true, only comparisons that compare a boolean literal to a boolean will be checked.
+  // if false, comparisons that compare a boolean literal to a nullable boolean variable will also be checked
+  allowComparingNullableBooleans?: boolean;
+};
+
+const defaults = {
+  allowComparingNullableBooleans: true,
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/no-unnecessary-boolean-literal-compare.md:51]:
```suggestion
  allowComparingNullableBooleans: false,
```

**Comment by zachkirsch** [comment]
Sweet! I actually am having some second thoughts haha. I feel like auto-fixing `varBooleanOrNull === false` to `!(varBooleanOrNull ?? true)` is kind of weird - the "fixed" version is also a bit hard to parse visually.

Fixing `varBooleanOrNull === true` to `varBooleanOrNull` and `varBooleanOrNull !== true` to `!varBooleanOrNull` is good IMO. I think fixing `varBooleanOrNull !== false` to `varBooleanOrNull ?? true` is also good but I could see it either way.

What are your thoughts? We could also have additional config options that more granularly control which of these four comparisons are checked/fixed by the rule.

File: packages/eslint-plugin/src/rules/no-unnecessary-boolean-literal-compare.ts (hunk 1 above)
Comment (zachkirsch) [reply packages/eslint-plugin/src/rules/no-unnecessary-boolean-literal-compare.ts:72]: Just committed these before realizing that the default before (`allowComparingNullableBooleans=true`) was backwards compatible, and switching it to `false` is no longer backwards compatible.

`allowComparingNullableBooleans=true` means "comparing literal booleans to nullable booleans is allowed so eslint should not complain if I do it" and `allowComparingNullableBooleans=false` means "comparing literal booleans to nullable booleans is not allowed so eslint should complain if I do it".

Happy to change the naming since clearly it's a bit confusing :)

**Comment by bradzacher** [comment]
Hmmm. Sorry, I did a quick eyeball and didn't notice the fix output.
That's my bad.

I think a fixer might be bad here.

- `foo === true` => `foo` - treat the false and undefined cases the same
- `foo !== true` => `!foo` - treat the false and undefined cases the same 
- `foo === false` => `foo ?? true` - 
- `foo !== false` => `!(foo ?? true)` - treat the t

**Comment by zachkirsch** [comment]
Okay, I broke `allowComparingNullableBooleans` into two options: `allowComparingNullableBooleansToTrue` and `allowComparingNullableBooleansToFalse`, so users can decide what they want to allow. If you want, I'm happy to just have the rule ignore comparisons between nullable boolean variables and literal `false`, but I think it might be confusing if eslint complains about `booleanOrNullVar === true` but not `booleanOrNullVar === false`.

I also added some more docs to the rule's readme, including a table of what exactly the fixer does in each case.

**Comment by zachkirsch** [comment]
Lemme know if you need anything else from me!
